### PR TITLE
[rllib] Add starcraft multiagent env as example

### PR DESCRIPTION
**Pull request by richardliaw**
## What do these changes do?

Based off https://github.com/oxwhirl/pymarl_alpha/, add a MultiAgentEnv example for SC2.

These basic comparison tests are run on a default map (`3m_3m`) with a max score of 20. The PPO and PyMARL scores are comparable.

## From Original Repository (PyMARL):

screenshot 2018-12-18 22 21 18
screenshot 2018-12-18 22 21 26

Replicate this by installing given pymarl instructions and running:
`python src/main.py --config=qmix --env-config=sc2`. 

## From RLlib using our connector (PPO):
This uses PPO with 4 workers to train FCNet on StarCraft 2.

Replicate this with the following commands (assuming PyMARL is installed)

```
export PYMARL_PATH=/path/to/pymarl
python starcraft_env.py --run PPO
```

screenshot 2018-12-21 00 00 33

The RLlib QMix does not achieve expected performance:

```
export PYMARL_PATH=/path/to/pymarl
python starcraft_env.py --run QMIX
```

screenshot 2018-12-21 00 42 15

## Related issue number
#3540

### CODE REVIEW

**Comment by richardliaw** [comment]
Running PPO with 2 workers: 

screenshot 2018-12-14 18 14 37
screenshot 2018-12-14 18 14 44

**Comment by richardliaw** [comment]
Running `python3 src/main.py --config=qmix --env-config=sc2 with env_args.map_name=2s_3z` in https://github.com/oxwhirl/pymarl_alpha

screenshot 2018-12-14 18 09 12
screenshot 2018-12-14 18 09 59
screenshot 2018-12-14 18 18 33

So I guess I'm doing something wrong since the episode length mean is way longer in this one.

**Comment by richardliaw** [comment]
Potential issues - not properly masking actions?

**Comment by ericl** [comment]
Is it possible to compare videos to see what's going on? Action masking shouldn't be hard if you change the obs space to a dict, then use: https://ray.readthedocs.io/en/latest/rllib-models.html#variable-length-parametric-action-spaces

(just use masking, not action embeddings).

**Comment by ericl** [comment]
Btw, rewards are summed in the main metric (looks like 5 agents in that scenario => 60/5 = 12?). You should look at the policy_reward metric to see what the mean reward for a particular policy is.

**Comment by richardliaw** [comment]
Right now RLlib is getting something like:

```
custom_metrics: {}
date: 2018-12-18_22-30-33
done: false
episode_len_mean: 21.71
episode_reward_max: 60.000000000000036
episode_reward_mean: 41.99342465753425
episode_reward_min: 16.931506849315063
episodes_this_iter: 10
episodes_total: 955
experiment_id: d42b9067823f49c69e4f49087cf35635
hostname: c73
info:
  grad_time_ms: 8.273
  learner: {}
  num_steps_sampled: 60000
  num_steps_trained: 60000
  opt_peak_throughput: 72522.525
  opt_samples: 600.0
  sample_peak_throughput: 402.19
  sample_time_ms: 1491.831
  update_time_ms: 0.002
iterations_since_restore: 100
node_ip: 169.229.49.173
num_metric_batches_dropped: 0
pid: 27303
policy_reward_mean: {}
time_since_restore: 154.1551558971405
time_this_iter_s: 1.564133644104004
time_total_s: 154.1551558971405
timestamp: 1545201033
timesteps_since_restore: 60000
timesteps_this_iter: 600
timesteps_total: 60000
training_iteration: 100
```

**Comment by richardliaw** [comment]
Why is it that PGAgent works fine, but PPOAgent throws this error (without changing any parameters, and with the same model)??: 

```python
WARN: gym.spaces.Box autodetected dtype as <class 'numpy.float32'>. Please provide explicit dtype.
WARN: gym.spaces.Box autodetected dtype as <class 'numpy.float32'>. Please provide explicit dtype.
2018-12-20 19:49:33.161022: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
Traceback (most recent call last):
  File "starcraft_env.py", line 150, in <module>
    agent = PPOAgent(env="starcraft", config=agent_cfg)
  File "/data/rliaw/miniconda3/envs/pymarl/lib/python3.6/site-packages/ray/rllib/agents/agent.py", line 244, in __init__
    Trainable.__init__(self, config, logger_creator)
  File "/data/rliaw/miniconda3/envs/pymarl/lib/python3.6/site-packages/ray/tune/trainable.py", line 87, in __init__
    self._setup(copy.deepcopy(self.config))
  File "/data/rliaw/miniconda3/envs/pymarl/lib/python3.6/site-packages/ray/rllib/agents/agent.py", line 312, in _setup
    self._init()
  File "/data/rliaw/miniconda3/envs/pymarl/lib/python3.6/site-packages/ray/rllib/agents/ppo/ppo.py", line 72, in _init
    self.env_creator, self._policy_graph)
  File "/data/rliaw/miniconda3/envs/pymarl/lib/python3.6/site-packages/ray/rllib/agents/agent.py", line 413, in make_local_evaluator
    config["local_evaluator_tf_session_args"]
  File "/data/rliaw/miniconda3/envs/pymarl/lib/python3.6/site-packages/ray/rllib/agents/agent.py", line 518, in _make_evaluator
    output_creator=output_creator)
  File "/data/rliaw/miniconda3/envs/pymarl/lib/python3.6/site-packages/ray/rllib/evaluation/policy_evaluator.py", line 267, in __init__
    self._build_policy_map(policy_dict, policy_config)
  File "/data/rliaw/miniconda3/envs/pymarl/lib/python3.6/site-packages/ray/rllib/evaluation/policy_evaluator.py", line 571, in _build_policy_map
    policy_map[name] = cls(obs_space, act_space, merged_conf)
  File "/data/rliaw/miniconda3/envs/pymarl/lib/python3.6/site-packages/ray/rllib/agents/ppo/ppo_policy_graph.py", line 199, in __init__
    }, observation_space, 1, vf_config).outputs
  File "/data/rliaw/miniconda3/envs/pymarl/lib/python3.6/site-packages/ray/rllib/models/catalog.py", line 195, in get_model
    options, state_in, seq_lens)
  File "/data/rliaw/miniconda3/envs/pymarl/lib/python3.6/site-packages/ray/rllib/models/catalog.py", line 224, in _get_model
    seq_lens=seq_lens)
  File "/data/rliaw/miniconda3/envs/pymarl/lib/python3.6/site-packages/ray/rllib/models/model.py", line 72, in __init__
    num_outputs, options)
  File "starcraft_env.py", line 30, in _build_layers_v2
    num_outputs, action_mask)
ValueError: ('This model assumes num outputs is equal to max avail actions', 1, <tf.Tensor 'default/value_function/Reshape:0' shape=(?, 9) dtype=float32>)
Exception ignored in: <bound method PolicyEvaluator.__del__ of <ray.rllib.evaluation.policy_evaluator.PolicyEvaluator object at 0x7f1a0938c668>>
Traceback (most recent call last):
  File "/data/rliaw/miniconda3/envs/pymarl/lib/python3.6/site-packages/ray/rllib/evaluation/policy_evaluator.py", line 575, in __del__
    if isinstance(self.sampler, AsyncSampler):
AttributeError: 'PolicyEvaluator' object has no attribute 'sampler'
RequestQuit command received.
```

**Comment by ericl** [comment]
PPO has those free log std outputs? But that is only for continuous action
spaces, this is discrete right?

**Comment by ericl** [comment]
ValueError: ('This model assumes num outputs is equal to max avail
actions', 1, <tf.Tensor 'default/value_function/Reshape:0' shape=(?, 9)
dtype=float32>

Oh I think you need to disable the value function. See the cartpole PPO
example.

**Comment by richardliaw** [comment]
this one? ray/rllib/examples/multiagent_cartpole.py 

I can't seem to find where there is indication of disabling value function:

```

if __name__ == "__main__":
    args = parser.parse_args()
    ray.init()

    # Simple environment with `num_agents` independent cartpole entities
    register_env("multi_cartpole", lambda _: MultiCartpole(args.num_agents))
    ModelCatalog.register_custom_model("model1", CustomModel1)
    ModelCatalog.register_custom_model("model2", CustomModel2)
    single_env = gym.make("CartPole-v0")
    obs_space = single_env.observation_space
    act_space = single_env.action_space

    # Each policy can have a different configuration (including custom model)
    def gen_policy(i):
        config = {
            "model": {
                "custom_model": ["model1", "model2"][i % 2],
            },
            "gamma": random.choice([0.95, 0.99]),
        }
        return (PPOPolicyGraph, obs_space, act_space, config)

    # Setup PPO with an ensemble of `num_policies` different policy graphs
    policy_graphs = {
        "policy_{}".format(i): gen_policy(i)
        for i in range(args.num_policies)
    }
    policy_ids = list(policy_graphs.keys())

    run_experiments({
        "test": {
            "run": "PPO",
            "env": "multi_cartpole",
            "stop": {
                "training_iteration": args.num_iters
            },
            "config": {
                "log_level": "DEBUG",
                "num_sgd_iter": 10,
                "multiagent": {
                    "policy_graphs": policy_graphs,
                    "policy_mapping_fn": tune.function(
                        lambda agent_id: random.choice(policy_ids)),
                },
            },
        }
    })

```

**Comment by richardliaw** [comment]
screenshot 2018-12-21 00 00 33

Looks pretty good now (4 workers PPO)

**Comment by richardliaw** [comment]
screenshot 2018-12-21 00 42 15

**Comment by ericl** [comment]
Can you update the PR description to summarize the perf numbers vs QMIX, and with exact commands on how they were produced?me Don'tgraphs you posted mean.

**Comment by richardliaw** [comment]
Updated

File: python/ray/rllib/examples/starcraft/twostep_game.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+import ray
+from ray.tune import run_experiments, function
+from ray.rllib.env.multi_agent_env import MultiAgentEnv
+from ray.rllib.agents.pg.pg_policy_graph import PGPolicyGraph
+from ray.rllib.agents.dqn.dqn_policy_graph import DQNPolicyGraph
+from ray.rllib.agents.ppo.ppo_policy_graph import PPOPolicyGraph
+
+
+class TwoStepGame(MultiAgentEnv):
```
Comment (ericl) [inline python/ray/rllib/examples/starcraft/twostep_game.py]: Remove this file

**Review by ericl — APPROVED** [review]
I'm fine merging this for now, but ideally connectors using the public APIs should be maintained with their source repos and not in rllib itself.

Also nice to see that independent PPO is competitive in this scenario! I guess it's expected that our Qmix port isn't quite the same since it's using default DQN hyperparameters which aren't tuned as the pymarl ones are. Though it does seem to get there with time.

File: python/ray/rllib/examples/starcraft/starcraft_env.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+from ray.tune.registry import register_env
+from ray.rllib.models import Model, ModelCatalog
+from ray.rllib.models.misc import normc_initializer
+from ray.rllib.agents.qmix import QMixAgent
+from ray.rllib.agents.pg import PGAgent
+from ray.rllib.agents.ppo import PPOAgent
+from ray.tune.logger import pretty_print
+
+
+class ParametricActionsModel(Model):
```
Comment (ericl) [inline python/ray/rllib/examples/starcraft/starcraft_env.py]: Nit: this is just masking and not parametric actions.

**Pull request merged**